Provide list of editable fields via OCS API

rullzer opened the pull request:
Fixes #7983

CC: @tobiasKaminsky

rullzer committed:
Provide list of editable fields via OCS API

Fixes #7983

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented on apps/provisioning_api/appinfo/routes.php:
```diff
@@ -44,6 +44,7 @@
 		['root' => '/cloud', 'name' => 'Users#addUser', 'url' => '/users', 'verb' => 'POST'],
 		['root' => '/cloud', 'name' => 'Users#getUser', 'url' => '/users/{userId}', 'verb' => 'GET'],
 		['root' => '/cloud', 'name' => 'Users#getCurrentUser', 'url' => '/user', 'verb' => 'GET'],
+		['root' => '/cloud', 'name' => 'Users#getEditableField', 'url' => '/user/fields', 'verb' => 'GET'],
```
Suggestions for a better route are welcome ;)

rullzer commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -292,6 +292,38 @@ protected function getUserData($userId) {
+
+		$permittedFields[] = 'password';
+		if ($this->config->getSystemValue('force_language', false) === false) {
+			$permittedFields[] = 'language';
```
should we then also list the available languages?

tobiasKaminsky commented:
@przybylski this is for you :-)

MorrisJobke commented:
@rullzer What about tests?

rullzer commented:
@MorrisJobke will do once we all agree this is the API we want :wink:

juliushaertl commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -292,6 +292,38 @@ protected function getUserData($userId) {
+			$permittedFields[] = AccountManager::PROPERTY_EMAIL;
+		}
+
+		$permittedFields[] = 'password';
```
Changing passwords is not always possible, like on some ldap setups.

rullzer committed:
Don't expose language and password as editable

* Password might not always be editable
  - Also chaging that will invalide apptokens etc which will directly
  lock out your phone
* Chaging the language also doesn't make that much sense to me currently

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Fix tests

* Add factory for the federatedfilesharing app

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Now with tests

rullzer commented:
Review time ;)
